docs/custom-components.qmd

```diff
@@ -0,0 +1,314 @@
+---
+title: "Custom Javascript components"
+---
+
+In this post, you will learn how to create a custom element and accompanying output binding in Shiny. This is useful if you want to create an output that is not currently in Shiny for your app or if you want to create a custom output for a package you are developing.
+
+:::{.callout-note}
+The code shown here is simplified to get the point across, but before you use it in your own app, you should make sure to add error handling and other features to make it robust.
+:::
+
+# The problem
+
+You found a new table library that you really want to use in your Shiny app. Problem is, there's no wrapper for it, currently. The library is [Tabulator](https://tabulator.info/) and it's a javascript library for making tables with data.
+
+# The solution
+
+To implement a custom tabulator element for your app, you'll need to write three things:
+
+1. A javascript script that renders the element on the client side using the Tabulator library
+2. An `output_tabulator()` function for placing the element in your app's UI
+3. A `render_tabulator()` decorator for passing table data to the javascript code rendering the element on the server side
+
+
+## The javascript code
+
+First things first, to use a custom Javascript library we need to write... some javascript.
+
+To do this we will create a new folder called `tabulator/` that has the following structure:
+
+```
+tabulator/
+  tabulator_esm.min.js
+  tabulator.min.css
+  tableComponent.js
+```
+
+Both `tabulator_esm.min.js` and `tabulator.min.css` are downloaded from [tabulator's website.](https://tabulator.info/docs/5.5/install#sources-download) `tableComponent.js` is the script that we will write that contains the code for rendering the table to our Shiny app.
+
+:::{.callout-note}
+The code in this article will be abbreviated to show the relevant parts. If you want to see the full code, see the [accompanying repo.](https://github.com/nstrayer/pyshiny-output-binding-example)
+:::
+
+
+To create an output binding in Shiny, we create a new instance of the `Shiny.OutputBinding` class.
+
+```javascript
+class TabulatorOutputBinding extends Shiny.OutputBinding {
+    // Find element to render in
+    find(scope) { ... }
+
+    // Render output element in the found element
+    renderValue(el, payload) { ... }
+}
+
+// Register the binding
+Shiny.outputBindings.register(
+  new TabulatorOutputBinding(),
+  "shiny-tabulator-output"
+);
+```
+
+This class has two methods that we need to implement: `find()` and `renderValue()`. The `find()` method is used to find the element that we want to render the table in. The `renderValue()` method is used to render the table in the element. After making that class we need to register it with Shiny so it can find and send data to instances of our output.
```

Comment:
```suggestion
This class has two methods that we need to implement: `find()` and `renderValue()`. The `find()` method is used to identify the element that will contain the rendered table. The `renderValue()` method is used to render the table in the element. After making that class we need to register it with Shiny so it can find and send data to instances of our output.
```